File: benchmarks/tsb/bench_timedelta_index_tostrings.ts
/**
 * Benchmark: TimedeltaIndex.toStrings(), .toArray(), .at(), .rename() on 10k-element index.
 * Outputs JSON: {"function": "timedelta_index_tostrings", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Timedelta, TimedeltaIndex } from "../../src/index.js";

const SIZE = 10_000;
const WARMUP = 5;
const ITERATIONS = 50;

const deltas = Array.from({ length: SIZE }, (_, i) =>
  Timedelta.fromComponents({ days: i % 365, hours: i % 24, minutes: i % 60 }),
);
const idx = TimedeltaIndex.fromTimedeltas(deltas, "duration");

for (let i = 0; i < WARMUP; i++) {
  idx.toStrings();
  idx.toArray();
  idx.at(0);
  idx.at(SIZE - 1);
  idx.rename("elapsed");
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  idx.toStrings();
  idx.toArray();
  idx.at(0);
  idx.at(SIZE - 1);
  idx.rename("elapsed");
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "timedelta_index_tostrings",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_timedelta_index_tostrings.py
"""Benchmark: TimedeltaIndex.astype(str), .to_numpy(), element access, rename
on 10k-element TimedeltaIndex."""
import pandas as pd
import numpy as np
import json
import time

SIZE = 10_000
WARMUP = 5
ITERATIONS = 50

deltas = pd.to_timedelta(
    [(i % 365) * 24 * 3600 + (i % 24) * 3600 + (i % 60) * 60 for i in range(SIZE)],
    unit="s",
)
idx = pd.TimedeltaIndex(deltas, name="duration")

for _ in range(WARMUP):
    idx.astype(str)
    idx.to_numpy()
    idx[0]
    idx[-1]
    idx.rename("elapsed")

start = time.perf_counter()
for _ in range(ITERATIONS):
    idx.astype(str)
    idx.to_numpy()
    idx[0]
    idx[-1]
    idx.rename("elapsed")
total = time.perf_counter() - start

print(json.dumps({
    "function": "timedelta_index_tostrings",
    "mean_ms": total / ITERATIONS * 1000,
    "iterations": ITERATIONS,
    "total_ms": total * 1000,
}))
